rosbridge_library/test/internal/test_message_conversion.py

```diff
@@ -1,29 +1,20 @@
 #!/usr/bin/env python3
 import unittest
 from base64 import standard_b64encode
-from io import BytesIO
 from json import dumps, loads
 
-import rospy
-import rostest
+import numpy as np
+from rclpy.serialization import deserialize_message, serialize_message
 from rosbridge_library.internal import message_conversion as c
 from rosbridge_library.internal import ros_loader
 
 
 class TestMessageConversion(unittest.TestCase):
-    def setUp(self):
-        rospy.init_node("test_message_conversion")
-
     def validate_instance(self, inst1):
         """Serializes and deserializes the inst to typecheck and ensure that
         instances are correct"""
-        inst1._check_types()
-        buff = BytesIO()
-        inst1.serialize(buff)
-        inst2 = type(inst1)()
-        inst2.deserialize(buff.getvalue())
+        inst2 = deserialize_message(serialize_message(inst1), type(inst1))
         self.assertEqual(inst1, inst2)
```

Comment:
Do you understand what this function is supposed to be testing? It looks like it's using only standard rclpy functionality. Is it somehow testing any logic from rosbridge_library?

Reply:
@jtbandes I don't remember well... but I'll re-check and summarize.
- `inst1` and `inst2` are the ROS msg types
- ROS1 has `MsgClass.serialize()` and `MsgClass.deserialize()`, but ROS2 doesn't, so I changed them.
- This is a helper function that checks:
    - `inst1` can serialized using ROS function
    - `inst1` can deserialized back using ROS function, it will assigned to `inst2`

Reply:
I don't think that the `validate_instance` is a test by its self. Don't the method names need to be prefixed with `test` in order to be run as test cases? See https://docs.pytest.org/en/6.2.x/goodpractices.html#conventions-for-python-test-discovery

This is just a helper method used for assertions for various test cases. See for example `test_time_msg_now` which calls this function.

I think the reason this method was changed was to update it to use `rclpy` instead of `rospy` methods as an update to being ROS2 compatible.

Reply:
Yep, I understand it is a helper method used for other test cases. I still don't understand what it's for though — it looks like it takes a ROS message instance and runs it through the ROS serialization/deserialization layer and asserts the new instance is considered equal. That looks like it's only testing the ROS de/serialization code, not any of the rosbridge code?

Reply:
This function tests whether `inst`, which is defined as `inst = ros_loader.get_message_instance(msgtype)`, is the instance of a class that supports ROS serialization/deserialization.

In other words, this tests whether `ros_loader.get_message_instance` can generate such a class.
So I think it can be said "it's testing rosbridge code".

Reply:
@jtbandes 
Take for example this code (see the line with the comment added `# <==== Original`):
```python
    def do_byte_test(self, data_value, msgtype):
        for msg in [{"data": data_value}]:
            inst = ros_loader.get_message_instance(msgtype)
            c.populate_instance(msg, inst)
            self.assertEqual(inst.data, bytes([data_value]))
            self.validate_instance(inst) # <==== Original
            # ...
```

If we wanted to assert the `inst` object is a ROS2 message in some other way we could do something like: 
```python
    def do_byte_test(self, data_value, msgtype):
        for msg in [{"data": data_value}]:
            inst = ros_loader.get_message_instance(msgtype)
            c.populate_instance(msg, inst)
            self.assertEqual(inst.data, bytes([data_value]))
            self.assertTrue(isinstance(inst, eval(msgtype))) # <==== Changed for example
            # ...
```

Replacing `self.validate_instance(inst)` with `self.assertTrue(isinstance(inst, eval(msgtype)))`. However that does not work and even if it did I don't think we want to have `eval` in the test cases.

Do you think that checking if `inst` is a valid ROS2 message that can be serialized & deserialized successfully is appropriate in this test? Since the method `populate_instance` mutates `inst` can we say that we are checking the behavior of `rosbridge_library` `message_conversion`?

Reply:
> Do you think that checking if `inst` is a valid ROS2 message that can be serialized & deserialized successfully is appropriate in this test?

Not really, because `inst = ros_loader.get_message_instance(msgtype)` and there is another test file called `test_ros_loader.py`. 🤷